gerritlab/main.py

```diff
@@ -126,92 +121,78 @@ def cancel_prev_pipelines(commit):
+    push_time = 0
+    get_mrs_time = 0
+    update_mr_time = 0
+    create_mr_time = 0
 
+    # New approach
+    # Single push w/ all branch updates.
+    # Then create or update MRs.
+
+    # Push commits to Change-Id-named branches
+    refs_to_push = ["{}:refs/heads/{}".format(
+        c.commit.hexsha, c.source_branch) for c in commits_data]
+    print(f"Pushing {refs_to_push}")
+    start = time.time()
+    remote.push(refspec=refs_to_push, force=True)        
+    push_time += time.time() - start
+    print("done pushing")
+    
     # Get existing MRs created off of the given branch.
+    # ~400ms
+    start = time.time()
     current_mrs = merge_request.get_all_merge_requests(remote, local_branch)
-    current_mr_chain = merge_request.get_merge_request_chain(current_mrs)
+    get_mrs_time += time.time() - start
 
-    # Update the existing MRs from the end of the MR dependency chain.
-    updated_commits = []
+    current_mrs_by_source_branch = { mr.source_branch: mr for mr in current_mrs }
+
+    new_mrs = []
     updated_mrs = []
-    for mr in reversed(current_mr_chain):
-        for c in commits_data:
-            if c.source_branch == mr.source_branch:
-                # Delay the update if the updated target branch doesn't exist,
-                # meaning a new MR will be created off of that target branch.
-                updated_target_branch_exists = "{}/{}".format(
-                    remote.name,
-                    c.target_branch) in [ref.name for ref in repo.references]
-                if not updated_target_branch_exists:
-                    continue
-                # Do nothing if the commit doesn't contain any new changes.
-                if can_skip_ci(c.commit, c.source_branch):
-                    continue
-                # Update the target branch of this MR.
-                title, desp = utils.get_msg_title_description(c.commit.message)
+    
+    for c in commits_data:
+        title, desp = utils.get_msg_title_description(c.commit.message)
+        
+        mr = current_mrs_by_source_branch.get(c.source_branch)
+        if mr:
+            # Update the target branch of this MR (and other info)
+            print(f"{c} is linked to {mr}")
+
+            if mr.source_branch != c.source_branch:
+                print("SOURCE BRANCH NEEDS UPDATING")
+            if mr.target_branch != c.target_branch:
+                print("TARGET BRANCH NEEDS UPDATING")
+            if mr._title != title:
+                print("TITLE NEEDS UPDATING")
+            if mr._description != desp.strip():
+                print("DESC NEEDS UPDATING")
+                print("Desc extracted from commit:")
+                print("'" + desp.strip() + "'")
+                print("Desc from MR:")
+                print("'" + mr._description + "'")
+
+            if mr.source_branch != c.source_branch or mr.target_branch != c.target_branch or mr._title != title or mr._description != desp.strip():
```

Comment:
Other than these things, the updated MR can have a new commit (an amended commit). This is what the `can_skip_ci()` was checking.

Reply:
If the HEAD of the source branch of an MR changes (which would happen if a commit is amended), the MR self-updates with the new commit information.  It doesn't require an explicit update action here.  Maybe I'm misunderstanding the scenario you're thinking of?

Reply:
Ah you are right. I forgot that an MR will self-update with a new commit.